### PR TITLE
Add yarn image variants

### PR DESCRIPTION
**Pull request by promag**

### CODE REVIEW

File: 4/onbuild-yarn/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# Workaround because `WORKDIR` always runs as root.
+RUN chown -R node /home/node
+
+# Change the user to avoid running as `root`.
+USER node
+
+# Install dependencies.
+ONBUILD COPY *.json ./
+ONBUILD RUN yarn --ignore-scripts
```
Comment (ruimarinho) [inline 4/onbuild-yarn/Dockerfile]: Do we still need this (--ignore-scripts)?

File: 4/onbuild-yarn/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,48 @@
+FROM mhart/alpine-node:base-4
+
+ENV YARN_VERSION v0.16.1
+
+# Add yarn.
+RUN apk --no-cache add ca-certificates openssl \
+  && update-ca-certificates \
+  && wget https://github.com/yarnpkg/yarn/releases/download/${YARN_VERSION}/yarn-${YARN_VERSION}.tar.gz \
+  && tar zxvf yarn-${YARN_VERSION}.tar.gz
```
Comment (ruimarinho) [inline 4/onbuild-yarn/Dockerfile]: Move to its own fancy directory?

File: 4/onbuild-yarn/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+ENV PATH /dist/bin/:$PATH
+
+# Add gcc and git support for native dependencies.
+RUN apk --no-cache --virtual build-dependencies add g++ gcc git make python
```
Comment (ruimarinho) [inline 4/onbuild-yarn/Dockerfile]: Move up since we'll update yarn far more than these dependencies?

File: 4/onbuild-yarn/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+ENV YARN_VERSION v0.16.1
+
+# Add yarn.
+RUN apk --no-cache add ca-certificates openssl \
+  && update-ca-certificates \
+  && wget https://github.com/yarnpkg/yarn/releases/download/${YARN_VERSION}/yarn-${YARN_VERSION}.tar.gz \
+  && tar zxvf yarn-${YARN_VERSION}.tar.gz
+
+ENV PATH /dist/bin/:$PATH
```
Comment (ruimarinho) [inline 4/onbuild-yarn/Dockerfile]: Move up?

**Review by ruimarinho — CHANGES_REQUESTED** [review]

File: 4/onbuild-yarn/Dockerfile (hunk 2 above)
Comment (promag) [reply 4/onbuild-yarn/Dockerfile]: Delete archive after `tar zx`.

File: 6/slim-yarn/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+# Add gcc and git support for native dependencies.
+RUN apk --no-cache --virtual build-dependencies add g++ gcc git make python
+
+# Add yarn.
+ARG YARN_VERSION=v0.16.1
+
+ENV PATH /usr/local/lib/yarn/bin/:$PATH
+
+RUN apk --no-cache --virtual yarn-dependencies add ca-certificates openssl \
```
Comment (ruimarinho) [inline 6/slim-yarn/Dockerfile]: add `tar` here so that a newly updated version is installed.

File: 6/slim-yarn/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  && update-ca-certificates \
+  && wget https://github.com/yarnpkg/yarn/releases/download/${YARN_VERSION}/yarn-${YARN_VERSION}.tar.gz \
```
Comment (ruimarinho) [inline 6/slim-yarn/Dockerfile]: `wget -O - https://github.com/yarnpkg/yarn/releases/download/${YARN_VERSION}/yarn-${YARN_VERSION}.tar.gz | tar -C /usr/local/lib/yarn/ --strip-components 1 -zxvf -`

**Review by ruimarinho — CHANGES_REQUESTED** [review]

**Comment by promag** [comment]
For reference, this is the diff between npm and yarn.

``` diff
diff 7/slim/Dockerfile 7/slim-yarn/Dockerfile
1c1
< FROM mhart/alpine-node:7
---
> FROM mhart/alpine-node:base-7
5a6,16
> # Add yarn.
> ARG YARN_VERSION=v0.16.1
>
> ENV PATH /usr/local/lib/yarn/bin/:$PATH
>
> RUN apk --no-cache --virtual yarn-dependencies add ca-certificates openssl tar \
>   && update-ca-certificates \
>   && mkdir /usr/local/lib/yarn/ \
>   && wget -O - https://github.com/yarnpkg/yarn/releases/download/${YARN_VERSION}/yarn-${YARN_VERSION}.tar.gz | tar -C /usr/local/lib/yarn/ --strip-components 1 -zxvf - \
>   && apk del yarn-dependencies
>
18,19c29,30
< # Copy package.json and npm-shrinkwrap.json if they exist.
< ONBUILD COPY *.json ./
---
> # Copy package.json and yarn.lock if they exist.
> ONBUILD COPY *package.json *yarn.lock ./
```

File: 7/test-yarn/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,21 @@
+FROM mhart/alpine-node:base-7
+
+# Add gcc and git support for native dependencies.
+RUN apk --no-cache --virtual build-dependencies add g++ gcc git make python
+
+# Add yarn.
+ARG YARN_VERSION v0.16.1
```
Comment (ruimarinho) [inline 7/test-yarn/Dockerfile]: Missing `=`.

**Review by ruimarinho — CHANGES_REQUESTED** [review]

File: 4/test-yarn/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,21 @@
+FROM mhart/alpine-node:base-4
+
+# Add gcc and git support for native dependencies.
+RUN apk --no-cache --virtual build-dependencies add g++ gcc git make python
+
+# Add yarn.
+ARG YARN_VERSION v0.16.1
```
Comment (ruimarinho) [inline 4/test-yarn/Dockerfile]: Missing `=`.

File: 5/test-yarn/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,21 @@
+FROM mhart/alpine-node:base-5
+
+# Add gcc and git support for native dependencies.
+RUN apk --no-cache --virtual build-dependencies add g++ gcc git make python
+
+# Add yarn.
+ARG YARN_VERSION v0.16.1
```
Comment (ruimarinho) [inline 5/test-yarn/Dockerfile]: Missing `=`.

**Review by ruimarinho — CHANGES_REQUESTED** [review]

**Review by ruimarinho — APPROVED** [review]

**Pull request merged**